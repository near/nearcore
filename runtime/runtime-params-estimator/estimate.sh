#!/usr/bin/env bash

./setup.sh

vmkind="wasmer"
features="required"

if [[ ! -z "$1" ]]; then
  features="$features,$1"
  if [[ "$1" == *"wasmtime"* || "$1" == *"lightbeam"* ]]; then
    vmkind="wasmtime";
  fi
fi

rm -rf /tmp/data

set -ex

script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"

cd "${script_dir}"/../../
cargo run --release --package neard --bin neard -- --home /tmp/data init --test-seed=alice.near --account-id=test.near --fast
cd "${script_dir}"/../../genesis-tools/genesis-populate
cargo run --release --package genesis-populate --bin genesis-populate -- --additional-accounts-num=10000000 --home /tmp/data
cd "${script_dir}"
cargo build --release --package runtime-params-estimator --features $features
./emu-cost/counter_plugin/qemu-x86_64 -cpu Westmere-v1 -plugin file=./emu-cost/counter_plugin/libcounter.so ../../target/release/runtime-params-estimator --home /tmp/data --accounts-num 1000000 --iters 1 --warmup-iters 0 --warmup-transactions 0 --vm-kind $vmkind

