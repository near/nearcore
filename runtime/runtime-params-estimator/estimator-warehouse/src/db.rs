//! Database connection and struct representing rows in the data tables.

use std::path::Path;

use chrono::NaiveDateTime;
use rusqlite::{params, Connection, Row};

use crate::Metric;

/// Wrapper around database connection
pub(crate) struct Db {
    conn: Connection,
}

impl Db {
    pub(crate) fn new(conn: Connection) -> Self {
        Self { conn }
    }

    /// Opens an existing SQLite Db or creates it
    pub(crate) fn open(path: &Path) -> anyhow::Result<Self> {
        let conn = Connection::open(path)?;
        let init_sql = include_str!("init.sql");
        conn.execute_batch(init_sql)?;
        Ok(Self::new(conn))
    }
}

/// A single data row in the estimation table
#[derive(Debug, PartialEq)]
pub(crate) struct EstimationRow {
    /// Name of the estimation / parameter
    pub name: String,
    /// The estimation result converted to gas units
    pub gas: f64,
    /// Parameter for which this estimation is used
    pub parameter: Option<String>,
    /// The estimated time in nanoseconds, (if time-based estimation)
    pub wall_clock_time: Option<f64>,
    /// The number of operations counted (if icount-based estimation)
    pub icount: Option<f64>,
    /// The number of IO read bytes counted (if icount-based estimation)
    pub io_read: Option<f64>,
    /// The number of IO write bytes counted (if icount-based estimation)
    pub io_write: Option<f64>,
    /// For measurements that had some kind of inaccuracies or problems
    pub uncertain_reason: Option<String>,
    /// Which git commit this has been estimated on
    pub commit_hash: String,
}

/// A single data row in the parameter table
#[derive(Debug, PartialEq)]
pub(crate) struct ParameterRow {
    /// Name of the estimation / parameter
    pub name: String,
    /// The estimation result converted to gas units
    pub gas: f64,
    /// Protocol version for which the parameter is valid
    pub protocol_version: u32,
}

impl EstimationRow {
    const SELECT_ALL: &'static str =
        "name,gas,parameter,wall_clock_time,icount,io_read,io_write,uncertain_reason,commit_hash";
    pub fn get(db: &Db, name: &str, commit: &str, metric: Metric) -> anyhow::Result<Vec<Self>> {
        Ok(Self::get_any_metric(db, name, commit)?
            .into_iter()
            .filter(|row| row.is_metric(metric))
            .collect())
    }
    pub fn get_any_metric(db: &Db, name: &str, commit: &str) -> anyhow::Result<Vec<Self>> {
        let select = Self::SELECT_ALL;
        let mut stmt = db.conn.prepare(&format!(
            "SELECT {select} FROM estimation WHERE name = ?1 AND commit_hash = ?2;"
        ))?;
        let data = stmt
            .query_map([name, commit], Self::from_row)?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }
    pub(crate) fn insert(&self, db: &Db) -> anyhow::Result<()> {
        db.conn.execute(
            "INSERT INTO estimation(name,gas,parameter,wall_clock_time,icount,io_read,io_write,uncertain_reason,commit_hash) values (?1,?2,?3,?4,?,?6,?7,?8,?9)",
            params![
                self.name,
                self.gas,
                self.parameter,
                self.wall_clock_time,
                self.icount,
                self.io_read,
                self.io_write,
                self.uncertain_reason,
                self.commit_hash,
            ],
        )?;
        Ok(())
    }
    pub fn select_by_commit_and_metric(
        db: &Db,
        commit: &str,
        metric: Metric,
    ) -> anyhow::Result<Vec<Self>> {
        let select = Self::SELECT_ALL;
        let metric_condition = metric.condition();
        let mut stmt = db.conn.prepare(&format!(
            "SELECT {select} FROM estimation WHERE commit_hash = ?1 AND {metric_condition};"
        ))?;
        let data = stmt
            .query_map([commit], Self::from_row)?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }

    /// Returns one (commit_hash,date) tuple for each commit in store,
    /// optionally filtered by estimation metric. The output is sorted by the
    /// date, in ascending order. Note that the date is not the committed-date
    /// but instead the earliest date associated with an estimation of that
    /// commit, as recorded in the warehouse. This is usually the date the
    /// estimation has been performed.
    pub fn commits_sorted_by_date(
        db: &Db,
        metric: Option<Metric>,
    ) -> anyhow::Result<Vec<(String, NaiveDateTime)>> {
        let metric_condition = match metric {
            Some(m) => format!("WHERE {}", m.condition()),
            None => "".to_owned(),
        };
        let sql = format!("SELECT commit_hash,min(date) FROM estimation {metric_condition} GROUP BY commit_hash ORDER BY date ASC;");
        let mut stmt = db.conn.prepare(&sql)?;
        let data = stmt
            .query_map([], |row| Ok((row.get::<_, String>(0)?, row.get::<_, NaiveDateTime>(1)?)))?
            .collect::<Result<Vec<_>, rusqlite::Error>>()?;
        Ok(data)
    }
    pub fn count_by_metric(db: &Db, metric: Metric) -> anyhow::Result<u64> {
        let sql = match metric {
            Metric::ICount => "SELECT COUNT(*) FROM estimation WHERE icount IS NOT NULL;",
            Metric::Time => "SELECT COUNT(*) FROM estimation WHERE wall_clock_time IS NOT NULL;",
        };
        let count = db.conn.query_row::<i32, _, _>(sql, [], |row| row.get(0))?;
        Ok(count as u64)
    }
    pub fn last_updated(db: &Db, metric: Metric) -> anyhow::Result<Option<NaiveDateTime>> {
        let sql = match metric {
            Metric::ICount => "SELECT MAX(date) FROM estimation WHERE icount IS NOT NULL;",
            Metric::Time => "SELECT MAX(date) FROM estimation WHERE wall_clock_time IS NOT NULL;",
        };
        let dt = db.conn.query_row::<Option<NaiveDateTime>, _, _>(sql, [], |row| row.get(0))?;
        Ok(dt)
    }
    fn is_metric(&self, metric: Metric) -> bool {
        match metric {
            Metric::ICount => self.icount.is_some(),
            Metric::Time => self.icount.is_none() && self.wall_clock_time.is_some(),
        }
    }
    fn from_row(row: &Row) -> rusqlite::Result<Self> {
        Ok(Self {
            name: row.get(0)?,
            gas: row.get(1)?,
            parameter: row.get(2)?,
            wall_clock_time: row.get(3)?,
            icount: row.get(4)?,
            io_read: row.get(5)?,
            io_write: row.get(6)?,
            uncertain_reason: row.get(7)?,
            commit_hash: row.get(8)?,
        })
    }
}

impl ParameterRow {
    pub fn count(db: &Db) -> anyhow::Result<u64> {
        let sql = "SELECT COUNT(*) FROM parameter;";
        let count = db.conn.query_row::<u64, _, _>(sql, [], |row| row.get(0))?;
        Ok(count)
    }
    pub fn latest_protocol_version(db: &Db) -> anyhow::Result<Option<u32>> {
        let sql = "SELECT MAX(protocol_version) FROM parameter;";
        let max = db.conn.query_row::<Option<u32>, _, _>(sql, [], |row| row.get(0))?;
        Ok(max)
    }
}

impl Metric {
    fn condition(&self) -> &'static str {
        match self {
            Metric::ICount => "icount IS NOT NULL",
            Metric::Time => "wall_clock_time IS NOT NULL",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Db;
    use crate::import::ImportConfig;
    use chrono::{NaiveDate, NaiveDateTime};
    use rusqlite::{functions::FunctionFlags, Connection};

    impl Db {
        /// Create a new in-memory test database with a mocked `datetime()` function.
        pub(crate) fn test() -> Self {
            let conn = Connection::open_in_memory().unwrap();
            let init_sql = include_str!("init.sql");
            conn.execute_batch(init_sql).unwrap();
            let db = Self::new(conn);
            db.mock_time(
                NaiveDate::from_ymd_opt(2015, 5, 15).unwrap().and_hms_opt(11, 22, 33).unwrap(),
            );
            db
        }

        /// Create a new in-memory test database with data defined by the input.
        ///
        /// The test data is expected to come in blocks of JSON lines with a commit header.
        /// WAIT statements can be used as barriers to make sure DB timestamps differ.
        pub(crate) fn test_with_data(input: &str) -> Self {
            let db = Self::test();
            for block in input.split("\n\n") {
                if block.trim() == "WAIT" {
                    // Update mocked time to ensure the following data is considered to be later.
                    let mocked_now: NaiveDateTime =
                        db.conn.query_row("SELECT datetime('now')", [], |r| r.get(0)).unwrap();
                    db.mock_time(mocked_now + chrono::Duration::seconds(1));
                } else {
                    let (commit_hash, input) = block.split_once("\n").unwrap();
                    let conf = ImportConfig {
                        commit_hash: Some(commit_hash.to_string()),
                        protocol_version: None,
                    };
                    db.import_json_lines(&conf, input).unwrap();
                }
            }
            db
        }

        pub(crate) fn mock_time(&self, dt: NaiveDateTime) {
            let string_rep = dt.format("%Y-%m-%d %H:%M:%S").to_string();
            self.conn
                .create_scalar_function("datetime", 1, FunctionFlags::SQLITE_UTF8, move |_ctx| {
                    Ok(string_rep.clone())
                })
                .unwrap();
        }
    }
}
