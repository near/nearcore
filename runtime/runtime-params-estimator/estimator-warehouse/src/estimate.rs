use crate::{db::Db, import::ImportConfig};
use nix::unistd::Uid;
use xshell::{cmd, Shell};

/// Additional information required for estimation.
#[derive(Debug, clap::Parser)]
pub(crate) struct EstimateConfig {
    /// Specify the directory of a different repository, if not estimating the current one.
    #[clap(long)]
    pub external_repo: Option<String>,
    /// Specify the directory for near state used by the estimator. Will use
    /// temporary directory if unspecified.
    #[clap(long)]
    pub home: Option<String>,
    /// Comma separated list of metrics to use in estimation.
    #[clap(long, default_value = "icount,time", value_parser(["icount", "time"]), use_value_delimiter = true)]
    pub metrics: Vec<String>,
}

pub(crate) fn run_estimation(db: &Db, config: &EstimateConfig) -> anyhow::Result<()> {
    let sh = Shell::new()?;

    let mut _maybe_tmp = None;

    let estimator_home = match &config.home {
        Some(home) => home,
        None => {
            _maybe_tmp = Some(tempfile::tempdir()?);
            _maybe_tmp.as_ref().unwrap().path().to_str().unwrap()
        }
    };

    if let Some(external_repo) = &config.external_repo {
        sh.change_dir(external_repo);
    }
    let git_root = cmd!(sh, "git rev-parse --show-toplevel").read()?;

    // Ensure full optimization
    let _env_guard_one = sh.push_env("CARGO_PROFILE_RELEASE_LTO", "fat");
    let _env_guard_two = sh.push_env("CARGO_PROFILE_RELEASE_CODEGEN_UNITS", "1");

    // Build estimator
    cmd!(sh, "cargo build --release -p runtime-params-estimator --features runtime-params-estimator/required").run()?;
    // Find binary, some users have CARGO_TARGET_DIR pointing to a custom target directory
    let estimator_binary = if let Ok(target_dir) = sh.var("CARGO_TARGET_DIR") {
        format!("{target_dir}/release/runtime-params-estimator")
    } else {
        format!("{git_root}/target/release/runtime-params-estimator")
    };

    // Actual estimations
    let output = cmd!(sh, "git rev-parse HEAD").output()?;
    let mut commit_hash = String::from_utf8_lossy(&output.stdout).to_string();
    commit_hash.pop(); // \n
    let iters = 5.to_string();
    let warmup_iters = 1.to_string();

    if config.metrics.iter().any(|m| m == "time") {
        let mut maybe_drop_cache = vec![];

        #[cfg(target_family = "unix")]
        if Uid::effective().is_root() {
            maybe_drop_cache.push("--drop-os-cache");
        }

        if maybe_drop_cache.is_empty() {
            eprintln!("Running as non-root, storage related costs might be inaccurate because OS caches cannot be dropped");
        };

        let estimation_output =
            cmd!(sh,
                "{estimator_binary} --iters {iters} --warmup-iters {warmup_iters} --json-output --home {estimator_home} {maybe_drop_cache...} --metric time"
            ).read()?;
        db.import_json_lines(
            &ImportConfig { commit_hash: Some(commit_hash.clone()), protocol_version: None },
            &estimation_output,
        )?;
    }

    if config.metrics.iter().any(|m| m == "icount") {
        let estimation_output =
            cmd!(sh,
                "{estimator_binary} --iters {iters} --warmup-iters {warmup_iters} --json-output --home {estimator_home} --metric icount --docker --full"
            ).read()?;
        db.import_json_lines(
            &ImportConfig { commit_hash: Some(commit_hash), protocol_version: None },
            &estimation_output,
        )?;
    }

    Ok(())
}
