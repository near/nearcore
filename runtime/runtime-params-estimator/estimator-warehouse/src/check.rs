use crate::db::{Db, EstimationRow};
use crate::zulip::{ZulipEndpoint, ZulipReport};
use crate::Metric;
use std::collections::BTreeSet;

#[derive(clap::Parser, Debug)]
pub(crate) struct CheckConfig {
    /// Send notifications from checks to specified stream.
    /// Notifications are sent iff stream or user is set.
    #[clap(long)]
    zulip_stream: Option<String>,
    /// Send notifications from checks to specified Zulip user ID.
    /// Notifications are sent iff stream or user is set.
    #[clap(long)]
    zulip_user: Option<u64>,
    /// Checks have to be done on one specific metric.
    #[clap(long, value_enum)]
    metric: Metric,
    /// First git commit hash used for comparisons, used as base to calculate
    /// the relative changes. If left unspecified, the two commits that were
    /// inserted most recently are compared.
    #[clap(long)]
    commit_before: Option<String>,
    /// Second git commit hash used for comparisons. If left unspecified, the
    /// two commits that were inserted most recently are compared.
    #[clap(long)]
    commit_after: Option<String>,
    /// Names of estimations that should be checked. Leave empty to perform
    /// comparison on all available estimations.
    #[clap(long)]
    estimations: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) enum Status {
    Ok = 0,
    Warn = 1,
    // Critical = 2,
}

#[derive(Debug, PartialEq)]
pub(crate) enum Notice {
    RelativeChange(RelativeChange),
    UncertainChange(UncertainChange),
}

#[derive(Debug, PartialEq)]
pub(crate) struct RelativeChange {
    pub estimation: String,
    pub before: f64,
    pub after: f64,
}

#[derive(Debug, PartialEq)]
pub(crate) struct UncertainChange {
    pub estimation: String,
    pub before: String,
    pub after: String,
}

pub(crate) fn check(db: &Db, config: &CheckConfig) -> anyhow::Result<()> {
    let report = create_report(db, config)?;

    // This is the check command output to observe directly in the terminal.
    for change in report.changes() {
        println!("{change:?}");
    }

    let zulip_receiver = {
        if let Some(user) = config.zulip_user {
            Some(ZulipEndpoint::to_user(user)?)
        } else if let Some(stream) = &config.zulip_stream {
            Some(ZulipEndpoint::to_stream(stream.clone())?)
        } else {
            None
        }
    };

    if let Some(zulip) = zulip_receiver {
        zulip.post(&report)?;
    }
    Ok(())
}

pub(crate) fn create_report(db: &Db, config: &CheckConfig) -> anyhow::Result<ZulipReport> {
    let (commit_after, commit_before) = match (&config.commit_after, &config.commit_before) {
        (Some(a), Some(b)) => (a.clone(), b.clone()),
        (None, None) => {
            let mut commits = EstimationRow::commits_sorted_by_date(db, Some(config.metric))?;
            if commits.len() < 2 {
                anyhow::bail!("need data for at least 2 commits to perform comparison");
            }
            (commits.pop().unwrap().0, commits.pop().unwrap().0)
        }
        _ => anyhow::bail!("you have to either specify both commits for comparison or neither"),
    };
    let estimations = if !config.estimations.is_empty() {
        config.estimations.clone()
    } else {
        let rows_a = EstimationRow::select_by_commit_and_metric(db, &commit_after, config.metric)?;
        let rows_b = EstimationRow::select_by_commit_and_metric(db, &commit_before, config.metric)?;
        let estimations_a = rows_a.into_iter().map(|row| row.name).collect::<BTreeSet<_>>();
        let estimations_b = rows_b.into_iter().map(|row| row.name).collect::<BTreeSet<_>>();
        estimations_a.intersection(&estimations_b).cloned().collect()
    };
    let warnings =
        estimation_changes(db, &estimations, &commit_before, &commit_after, 0.1, config.metric)?;

    let warnings_uncertain = estimation_uncertain_changes(
        db,
        &estimations,
        &commit_before,
        &commit_after,
        config.metric,
    )?;

    let mut report = ZulipReport::new(commit_before, commit_after);
    for warning in warnings {
        report.add(warning, Status::Warn)
    }
    for warning in warnings_uncertain {
        report.add(warning, Status::Warn)
    }
    Ok(report)
}

fn estimation_changes(
    db: &Db,
    estimation_names: &[String],
    commit_before: &str,
    commit_after: &str,
    tolerance: f64,
    metric: Metric,
) -> anyhow::Result<Vec<Notice>> {
    let mut warnings = Vec::new();
    for name in estimation_names {
        let b = &EstimationRow::get(db, name, commit_before, metric)?[0];
        let a = &EstimationRow::get(db, name, commit_after, metric)?[0];
        let rel_change = (b.gas - a.gas).abs() / b.gas;
        if rel_change > tolerance {
            warnings.push(Notice::RelativeChange(RelativeChange {
                estimation: name.clone(),
                before: b.gas,
                after: a.gas,
            }))
        }
    }

    Ok(warnings)
}

fn estimation_uncertain_changes(
    db: &Db,
    estimation_names: &[String],
    commit_before: &str,
    commit_after: &str,
    metric: Metric,
) -> anyhow::Result<Vec<Notice>> {
    let mut warnings = Vec::new();
    for name in estimation_names {
        let b = EstimationRow::get(db, name, commit_before, metric)?.remove(0);
        let a = EstimationRow::get(db, name, commit_after, metric)?.remove(0);
        match (b.uncertain_reason, a.uncertain_reason) {
            (None, None) => continue,
            (Some(uncertain_before), None) => {
                add_warning(&mut warnings, name.clone(), uncertain_before, "None".to_owned())
            }
            (None, Some(uncertain_after)) => {
                add_warning(&mut warnings, name.clone(), "None".to_owned(), uncertain_after)
            }
            (Some(uncertain_before), Some(uncertain_after)) => {
                add_warning(&mut warnings, name.clone(), uncertain_before, uncertain_after);
            }
        }
    }

    Ok(warnings)
}

fn add_warning(warnings: &mut Vec<Notice>, name: String, before: String, after: String) {
    warnings.push(Notice::UncertainChange(UncertainChange { estimation: name, before, after }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[track_caller]
    fn generate_test_report(input: &str, metric: Metric, estimations: &[&str]) -> ZulipReport {
        let db = Db::test_with_data(input);
        let config = CheckConfig {
            zulip_stream: None,
            zulip_user: None,
            metric,
            commit_before: None,
            commit_after: None,
            estimations: estimations.iter().map(|&s| s.to_owned()).collect(),
        };
        create_report(&db, &config).unwrap()
    }

    #[test]
    fn test_check_command() {
        let input_a = r#"
        0000a
        {"computed_in":{"nanos":800,"secs":44},"name":"LogBase","result":{"gas":1000000000.0,"instructions":8000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":809,"secs":26},"name":"LogByte","result":{"gas":1000000000.0,"instructions":8000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0001a
        {"computed_in":{"nanos":814,"secs":9},"name":"LogBase","result":{"gas":2000000000.0,"instructions":16000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":694,"secs":33},"name":"LogByte","result":{"gas":1002000000.0,"instructions":8016.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0002a
        {"computed_in":{"nanos":331,"secs":24},"name":"LogBase","result":{"gas":3000000000.0,"instructions":24000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":511,"secs":52},"name":"LogByte","result":{"gas":1004000000.0,"instructions":8032.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        0003a
        {"computed_in":{"nanos":633,"secs":7},"name":"LogBase","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":173,"secs":2},"name":"LogByte","result":{"gas":1006000000.0,"instructions":8048.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"NEGATIVE-COST"}}
        {"computed_in":{"nanos":655,"secs":56},"name":"LogByte","result":{"gas":20000000.0,"time_ns":20,"metric":"time","uncertain_reason":null}}

        0004a
        {"computed_in":{"nanos":319,"secs":19},"name":"LogBase","result":{"gas":5000000000.0,"instructions":40000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":527,"secs":15},"name":"LogByte","result":{"gas":1008000000.0,"instructions":8064.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"HIGH-VARIANCE"}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest2","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":661,"secs":11},"name":"LogByte","result":{"gas":15000000.0,"time_ns":15,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":661,"secs":11},"name":"LogByte","result":{"gas":15000000.0,"time_ns":15,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128Sum","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128MultiExp","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        "#;

        // Only "LogBase" changes enough to show up in report.
        let report = generate_test_report(input_a, Metric::ICount, &[]);
        insta::assert_snapshot!(report.to_string());

        // Add more data and verify the notifications are updated.
        let input_b = input_a.to_owned()
            + r#"

        WAIT

        0000b
        {"computed_in":{"nanos":119,"secs":46},"name":"LogBase","result":{"gas":6000000000.0,"instructions":48000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":372,"secs":12},"name":"LogByte","result":{"gas":7000000000.0,"instructions":56000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":633,"secs":7},"name":"UncertainTest2","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":"BLOCK-MEASUREMENT-OVERHEAD"}}
        {"computed_in":{"nanos":262,"secs":15},"name":"LogByte","result":{"gas":20000000.0,"time_ns":20,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128Sum","result":{"gas":0.0,"time_ns":0,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":0,"secs":0},"name":"AltBn128MultiExp","result":{"gas":10.0,"time_ns":10,"metric":"time","uncertain_reason":null}}
        "#;

        // Now both estimations have changed.
        let report = generate_test_report(&input_b, Metric::ICount, &[]);
        insta::assert_snapshot!(report.to_string());

        // Verify that filter for specific estimations works.
        let report = generate_test_report(&input_b, Metric::ICount, &["LogBase"]);
        insta::assert_snapshot!(report.to_string());

        // Filter for metric.
        let report = generate_test_report(&input_b, Metric::Time, &[]);
        insta::assert_snapshot!(report.to_string());
    }
}
