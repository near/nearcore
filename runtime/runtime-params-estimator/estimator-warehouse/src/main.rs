use check::{check, CheckConfig};
use db::{Db, EstimationRow, ParameterRow};
use estimate::{run_estimation, EstimateConfig};
use import::ImportConfig;
use std::fmt::Write;
use std::io::{self, Read};
use std::path::PathBuf;

mod check;
mod db;
mod estimate;
mod import;
mod zulip;

#[derive(clap::Parser)]
struct CliArgs {
    #[clap(subcommand)]
    cmd: SubCommand,
    /// File path for either an existing SQLite3 DB or the path where a new DB
    /// will be created.
    #[clap(long, default_value = "db.sqlite")]
    db: PathBuf,
}

#[derive(clap::Subcommand, Debug)]
enum SubCommand {
    /// Call runtime-params-estimator for all metrics and import the results.
    Estimate(EstimateConfig),
    /// Read estimations in JSON format from STDIN and store it in the warehouse.
    Import(ImportConfig),
    /// Compares parameters, estimations, and how estimations changed over time.
    /// Reports any deviations from the norm to STDOUT. Combine with `--zulip`
    /// to send notifications to a Zulip stream
    Check(CheckConfig),
    /// Prints a summary of the current data in the warehouse.
    Stats,
}

fn main() -> anyhow::Result<()> {
    let cli_args: CliArgs = clap::Parser::parse();
    let db = Db::open(&cli_args.db)?;

    match cli_args.cmd {
        SubCommand::Estimate(config) => {
            run_estimation(&db, &config)?;
        }
        SubCommand::Import(config) => {
            let mut buf = String::new();
            io::stdin().read_to_string(&mut buf)?;
            db.import_json_lines(&config, &buf)?;
        }
        SubCommand::Check(config) => {
            check(&db, &config)?;
        }
        SubCommand::Stats => {
            let stats = generate_stats(&db)?;
            eprintln!("{stats}");
        }
    }

    Ok(())
}

#[derive(Debug, Clone, Copy, PartialEq, clap::ValueEnum)]
enum Metric {
    #[clap(name = "icount")]
    ICount,
    Time,
}

fn generate_stats(db: &Db) -> anyhow::Result<String> {
    let mut buf = String::new();
    writeln!(&mut buf)?;
    writeln!(&mut buf, "{:=^72}", " Warehouse statistics ")?;
    writeln!(&mut buf)?;
    writeln!(&mut buf, "{:>24}{:>24}{:>24}", "metric", "records", "last updated")?;
    writeln!(&mut buf, "{:>24}{:>24}{:>24}", "------", "-------", "------------")?;
    writeln!(
        &mut buf,
        "{:>24}{:>24}{:>24}",
        "icount",
        EstimationRow::count_by_metric(db, Metric::ICount)?,
        EstimationRow::last_updated(db, Metric::ICount)?
            .map(|dt| dt.to_string())
            .as_deref()
            .unwrap_or("never")
    )?;
    writeln!(
        &mut buf,
        "{:>24}{:>24}{:>24}",
        "time",
        EstimationRow::count_by_metric(db, Metric::Time)?,
        EstimationRow::last_updated(db, Metric::Time)?
            .map(|dt| dt.to_string())
            .as_deref()
            .unwrap_or("never")
    )?;
    writeln!(
        &mut buf,
        "{:>24}{:>24}{:>24}",
        "parameter",
        ParameterRow::count(db)?,
        ParameterRow::latest_protocol_version(db)?
            .map(|version| format!("v{version}"))
            .as_deref()
            .unwrap_or("never")
    )?;
    writeln!(&mut buf)?;
    writeln!(&mut buf, "{:=^72}", " END STATS ")?;

    Ok(buf)
}

#[cfg(test)]
mod tests {
    use super::generate_stats;
    use crate::db::Db;

    #[test]
    fn test_stats() {
        let input = r#"
        ae7f1cd2
        {"computed_in":{"nanos":222,"secs":1},"name":"LogBase","result":{"gas":1002000000.0,"time_ns":1002,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":671,"secs":18},"name":"LogBase","result":{"gas":1003000000.0,"instructions":8024.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        be7f1cd2
        {"computed_in":{"nanos":38,"secs":16},"name":"LogBase","result":{"gas":4000000000.0,"instructions":32000.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}
        {"computed_in":{"nanos":38,"secs":16},"name":"LogBase","result":{"gas":4000000000.0,"time_ns":4000,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":637,"secs":17},"name":"LogByte","result":{"gas":20100000.0,"instructions":160.8,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}

        ce7f1cd2
        {"computed_in":{"nanos":988,"secs":51},"name":"LogBase","result":{"gas":10000000000.0,"time_ns":10000,"metric":"time","uncertain_reason":null}}
        {"computed_in":{"nanos":441,"secs":54},"name":"LogByte","result":{"gas":20000000.0,"instructions":160.0,"io_r_bytes":0.0,"io_w_bytes":0.0,"metric":"icount","uncertain_reason":null}}"#;

        let db = Db::test_with_data(input);
        insta::assert_snapshot!(generate_stats(&db).unwrap());
    }
}
