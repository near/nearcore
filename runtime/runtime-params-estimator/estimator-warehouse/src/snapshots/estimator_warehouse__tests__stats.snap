---
source: runtime/runtime-params-estimator/estimator-warehouse/src/main.rs
expression: generate_stats(&db).unwrap()
---

========================= Warehouse statistics =========================

                  metric                 records            last updated
                  ------                 -------            ------------
                  icount                       4     2015-05-15 11:22:33
                    time                       3     2015-05-15 11:22:33
               parameter                       0                   never

============================== END STATS ===============================

